### ISSUE TITLE
How to use multi GPU

### ISSUE THREAD

**Issue by dh-Kang** [m-bain/whisperX#391]
Hi. sometimes, I need to transcribe recorded files that are tens of seconds to a few minutes long.
I used whisperX, but it takes long time for just transcribe. (I don't use align and diarization) It normally takes about half day.
I have 8GPUs, but whisperX use only one GPU while it is running.
I could find `v3` branch support multi GPU and `v3` branch is merged into `main` branch.
How can I use full GPUs?

```
whisperx.load_model('large-v2', 'cuda', compute_type='float16', device_index=list(range(8)), language="ko")
```
I load model with above code. And I tried to use batch size 16 and 64. 
But, Always, whisperX use only any one of 8 GPUs

**Comment by chenrq2005** [comment]
seems like the load_model function can only handle single int as device_index, but the fasterWhisper object handle int list when the fasterWhisper model being defined.

**Comment by chenrq2005** [comment]
in transcribe.py I modified these lines, seems like it worked, doing more verification.
26:     parser.add_argument("--device_index", default=0, type=Union[int, List[int]], help="device index to use for FasterWhisper inference")
86:     device_index: Union[int, List[int]] = args.pop("device_index")

**Comment by kyleboddy** [comment]
As far as I can tell using multi-GPU affords zero speedup. The code appears to just round-robin access the GPUs after loading the model into each of them.

**Comment by kyleboddy** [comment]
![GPU-toggle](https://github.com/m-bain/whisperX/assets/746351/a7b745cc-9395-492c-baba-dbcbe889900b)

Watch this GIF of `nvtop` being used. The GPUs are just used in round robin fashion, not concurrently. (Only my P40s are visible to the machine)

**Comment by chenrq2005** [comment]
Thanks @kyleboddy, my investigation told me the model will not process a single audio file with multiple GPUs concurrently.

**Comment by kyleboddy** [comment]
It could do this via chunking and splitting, then combining; there's logic out there that does it. But I don't believe this library handles it natively. Could make for a good pull request in the future.

**Comment by kyleboddy** [comment]
After running it with a single GPU and comparing, the multiGPU method using DEVICE_INDEX_LIST or similar is definitely used but only reduces memory pressure (by a good amount though, nearly half). It does not afford any computational speedup in its native setting.

**Comment by m-bain** [comment]
Thanks for the investigation kyle, yes it should be possible to shard to the audio segments over multiple GPUs and transcribe/align in parallel, will note this as future work

**Comment by supratim1121992** [comment]
I am using the WhisperX model on an EC2 instance with 8 V100 GPUs (16 GB each) and trying to process multiple files asynchronously using the ThreadPoolExecutor. The process runs fine but over time (after successfully processing over 500-1000 files) results in an error erratically with two different error messages:
1. CUDA error: misaligned address
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
2. CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

I have already tried looking into these errors and performed the following:

- Ensured that driver, cuda toolkit and cuDNN versions are updated to the current versions and compatible. I do however get this warning:
Model was trained with pyannote.audio 0.0.1, yours is 2.1.1. Bad things might happen unless you revert pyannote.audio to 0.x.
Model was trained with torch 1.10.0+cu102, yours is 2.0.1+cu117. Bad things might happen unless you revert torch to 1.x.
This is the `nvidia-smi` output:
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla V100-SXM2-16GB           Off | 00000000:00:17.0 Off |                    0 |
| N/A   29C    P0              57W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2-16GB           Off | 00000000:00:18.0 Off |                    0 |
| N/A   30C    P0              56W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2-16GB           Off | 00000000:00:19.0 Off |                    0 |
| N/A   29C    P0              58W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2-16GB           Off | 00000000:00:1A.0 Off |                    0 |
| N/A   30C    P0              56W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   4  Tesla V100-SXM2-16GB           Off | 00000000:00:1B.0 Off |                    0 |
| N/A   29C    P0              56W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   5  Tesla V100-SXM2-16GB           Off | 00000000:00:1C.0 Off |                    0 |
| N/A   30C    P0              54W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   6  Tesla V100-SXM2-16GB           Off | 00000000:00:1D.0 Off |                    0 |
| N/A   31C    P0              55W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   7  Tesla V100-SXM2-16GB           Off | 00000000:00:1E.0 Off |                    0 |
| N/A   29C    P0              55W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
- Monitored the GPU utilization, ensured that the memory usage does not exceed 16 GB and tuned the `batch_size` hyperparameter to 8, 16 & 32. The memory usage only goes upto 50% max but the GPU compute utilization goes upto 100% on multiple GPUs (due to multiple files being processed via multithreading).
-  Used `torch.compile` and `torch.nn.DataParallel` to parallelise the model but this still resulted in the error after processing over 500 files successfully.

This is how I am loading the WhisperX model and using it:
`input_asr_options =  {"beam_size": 3,"best_of": 3,"patience": 1,"length_penalty": 1,
                      "temperatures": [0.0, 0.2, 0.4, 0.6, 0.8, 1.0],"compression_ratio_threshold": 2.4,
                      "log_prob_threshold": -1.0,"no_speech_threshold": 0.6,"condition_on_previous_text": False,
                      "initial_prompt": None,"prefix": None,"suppress_blank": True,"suppress_tokens": [-1],
                      "without_timestamps": True,"max_initial_timestamp": 0.0,"word_timestamps": False,
                      "prepend_punctuations": "\"'“¿([{-","append_punctuations": "\"'.。,，!！?？:：”)]}、",
                      "suppress_numerals": False}
mod_trans = whisperx.load_model("medium.en", device = "cuda", device_index = [0,1,2,3,4,5,6,7], compute_type = "float16",
                                language = "en", asr_options = input_asr_options)
#mod_trans.transcribe = torch.compile(mod_trans.module.transcribe, mode = "max-autotune", backend = "inductor")
#mod_trans = torch.nn.DataParallel(mod_trans)

def download_transcribe(s3_key):
        proc_time = datetime.datetime.now().strftime("%Y-%m-%d_%H_%M_%S")
        try:
            with tempfile.NamedTemporaryFile(suffix = "." + fl_format, delete = True) as temp_in:
                # Downloading the file from S3 bucket
                s3_con.download_file(s3_in, s3_key, temp_in.name)
                # Loading the file and transcribing
                whisper_aud = whisperx.load_audio(temp_in.name)
                whisper_results = mod_trans.transcribe(whisper_aud, batch_size=32, num_workers = 1, language = "en")
                text = re.sub(r'\s+', ' ', "".join(seg["text"] for seg in whisper_results["segments"])).strip()
            return(text)

        except Exception as e:
            logging.error("Error encountered in file download and transcription: %s", e)
            return "Error"

# Function to perform multithreading using the above function with multiple workers (upto 8)
def transcribe_multi(s3_keys):
        with concurrent.futures.ThreadPoolExecutor(max_workers=num_gpu) as executor:
            futures = [executor.submit(download_transcribe, key) for key in s3_keys]
            results = [future.result() for future in concurrent.futures.as_completed(futures)]
        logging.info("Call transcription completed")
        return results
`

Any help or suggestions would be appreciated as I can't really get to the root cause of the error.

**Comment by kyleboddy** [comment]
I've been using ThreadPoolExecutor() as well to run on 3x Tesla P40 GPUs and still run into `CUDA error: an illegal memory access was encountered` every so often. It almost always occurs when I accidentally load a model to a GPU that has concurrent processing on it. I've added multiple GPU memory checks using GPUtil and other methods to ensure no overlapping, and that seems to solve the problem to like 99.999% so far.

Only suggestion is to add graceful error handling and restarting the state of the application, keeping track of the files using a locking semaphore or similar and reprocess after failure. Never got to the bottom of the actual issue.

**Comment by supratim1121992** [comment]
I'm loading the model on all 8 GPUs using `device_index` as can be seen in the code I shared. I'm not loading the files on any specific GPU for processing. The multiple threads created are being processed by the GPUs which are not utilised at the moment in my case. After quite a bit of tinkering and experimentation, I was finally able to resolve this error by upgrading my Pytorch version to the latest nightly release with Cuda 12.1 and the cuDNN version to 8.9.4 after referring to the support matrix available on Nvidia's webpage. I have been running the process over the past few days and have already processed over 50k audio files successfully without the error recurring even once. Hope that helps in resolving the issue in your case as well.

**Comment by DigilConfianz** [comment]
@kyleboddy 
I think threadpool executer failing is because the vad model is loaded on the wrong device(just cuda, with no device number mentioned.)
rewriting the https://github.com/m-bain/whisperX/blob/e94b9043085c32c365b2b60f23e73b2d03c2241c/whisperx/asr.py#L25C11-L25C11

https://github.com/m-bain/whisperX/blob/e94b9043085c32c365b2b60f23e73b2d03c2241c/whisperx/asr.py#L105

to
vad_model = load_vad_model(torch.device(f"{device}:{device_index}"), use_auth_token=None, **default_vad_options)
and passing the device as "cuda" and device_index to the device number helped me solve the issue.

**Comment by jamesmcarthur115555** [comment]
Csn you put the code into docker container , then launch 8 containers passing a different GPU device index to each container. The code inside container only sees one device. This would possibly guarantee each device is used.  If you are using less memory ina. Particular device you could maybe use MIG ifnavsiosbke to further partition the gpus

**Comment by lazyseacow** [comment]
@kyleboddy Thank you for your answer that helped me to solve most of the doubts. Have you ever tried to create multiple models with different GPU device numbers for each model and then use multithreading for concurrency? I'm not sure if this is feasible.

**Comment by kyleboddy** [comment]
I have not. I don't require that speed up as I'm running on Tesla P40s as-is; I'd switch to newer hardware first before trying to optimize multithreading. Sorry! IMO it is not possible to get a speed up using multiple GPUs.

**Comment by ewwink** [comment]
by using `device_index=[0,1]` it does use multiple GPU but it not well optimized. for optimized multi GPU/TPU support see [Whisper-Jax](https://github.com/sanchit-gandhi/whisper-jax/) but the downside, the model twice larger then faster-whisper so it maybe slower up to 100%
